<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Bregman-divergence based transfer-learning with application to trans-ethnic polygenic risk score • BRIGHT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Bregman-divergence based transfer-learning with application to trans-ethnic polygenic risk score">
<meta property="og:description" content="Polygenic risk scores (PRS) are a group of methods that solves the genetics risk prediction problem. However, single-ethnic PRS methods have limited performance on minority populations due to their limited sample sizes. Therefore, borrowing information from widely studied Caucasian populations can boost the prediction performance of minority models. In this paper, we efficiently implement the well-known high-dimensional transfer-learning method, prior LASSO, in the context of PRS construction with individual-level minority data and summary-level Caucasian data. We further extend it to a novel BRegman divergence-based Integrational Genetic Hazard Trans-ethnic (BRIGHT) method, which only requires GWAS summary data from both populations in order to preserve patient privacy and facilitate easier data sharing between institutes. In addition, we generalized the original LASSO penalty to a general penalty structure including group LASSO, group SCAD, and group MCP, which enable the group selection of important genetic markers while incorporating the LASSO, SCAD, and MCP penalties as special cases.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">BRIGHT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1"><div class="page-header"><h1 id="bright">BRIGHT<a class="anchor" aria-label="anchor" href="#bright"></a>
</h1></div></div>
<div class="section level1">
<h1 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h1>
<p><code>BRIGHT</code> is a group of methods for using individual-level data, published genotype-trait summary statistics (GWAS), or combined data types from different ethnic populations to improve the recalibration, discrimination, and prediction accuracy on the target minority cohort. We implemented group LASSO, Elastic Net, MCP and SCAD penalties for marker fine mapping.</p>
</div>
<div class="section level1">
<h1 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h1>
<p>Li, Q., Patrick, M. T., Zhang, H., Khunsriraksakul, C., Stuart, P. E., Gudjonsson, J. E., … &amp; He, K. (2022). Bregman Divergence-Based Data Integration with Application to Polygenic Risk Score (PRS) Heterogeneity Adjustment. arXiv preprint arXiv:2210.06025 (<a href="https://arxiv.org/abs/2210.06025" class="external-link uri">https://arxiv.org/abs/2210.06025</a>)</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p><code>BRIGHT</code> requires the following R packages: <code>Rcpp</code>, <code>Matrix</code>, <code>mvtnorm</code>, <code>BEDMatrix</code>, <code>grpreg</code>. Install them by:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Rcpp"</span>, <span class="st">"Matrix"</span>, <span class="st">"mvtnorm"</span>, <span class="st">"BEDMatrix"</span>, <span class="st">"grpreg"</span><span class="op">)</span>, dependencies<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>For Windows and Mac users, it would be easiest to download the following binaries (<a href="To%20be%20determined">Windows</a>, <a href="To%20be%20determined">Mac</a>) and install using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"/path/to/downloaded_binary_file"</span>, repos<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p>If you are on Linux or you would like to compile from source, you can download the source codes <a href="To%20be%20determined">BRIGHT_0.0.1.tar.gz</a>. Mac users should refer to <a href="https://cran.r-project.org/bin/macosx/tools/" class="external-link">this page</a> for the various dependencies required. Install then via:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"/path/to/downloaded_source.tar.gz"</span>, repos<span class="op">=</span><span class="cn">NULL</span>, type<span class="op">=</span><span class="st">"source"</span><span class="op">)</span></span></code></pre></div>
<p>If you have <code>devtools</code>, you can also type:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install_github</span><span class="op">(</span><span class="st">"To be determined"</span><span class="op">)</span></span></code></pre></div>
<p>or</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install_github</span><span class="op">(</span><span class="st">"To be determined"</span><span class="op">)</span></span></code></pre></div>
<p>for the latest development version. Or you can clone the latest development version here and install yourself using <code>devtools</code>.</p>
</div>
<div class="section level1">
<h1 id="warning">Warning!<a class="anchor" aria-label="anchor" href="#warning"></a>
</h1>
<p>Most functions in <code>BRIGHT</code> impute missing genotypes in PLINK bfiles with a homozygous A2 genotype, which is the same as using the <code>--fill-missing-a2</code> option in PLINK. It is the user’s responsibility to filter out individuals and SNPs with too many missing genotypes beforehand.</p>
</div>
<div class="section level1">
<h1 id="brights-tutorial">BRIGHTs tutorial<a class="anchor" aria-label="anchor" href="#brights-tutorial"></a>
</h1>
<p>BRIGHTs group of methods utilize a wide variety of summary-level data from different populations to carry out transfer-learning. We accounted for Linkage Disequilibrium (LD) via a reference panel (1000 genome project as default). The reference panel is assumed to be in PLINK 1 <a href="https://www.cog-genomics.org/plink/1.9/input#bed" class="external-link">format</a>. Summary statistics are expected to be loaded into memory as a data.frame/data.table.</p>
<p>Below we discuss the required data and implementation tutorials separately for quantitative traits and binary traits.</p>
<div class="section level2">
<h2 id="brights-with-quantitative-traits">BRIGHTs with quantitative traits<a class="anchor" aria-label="anchor" href="#brights-with-quantitative-traits"></a>
</h2>
<p>For quantitative traits, BRIGHTs requires the GWAS summary statistics or marginal genotype-trait inner product, <span class="math inline">$\frac{\boldsymbol X^\top\boldsymbol y}{n}$</span>, from the target minority population, while from the prior majority populations either GWAS summary statistics, marginal genotype-trait inner product, or coefficients estimated from joint models (e.g. PRS or LASSO regression) can be used for model fitting. We note that more than 1 prior majority data can be incorporated in the BRIGHTs model.</p>
<p>First we read the minority summary statistics and majority summary statistics into R, and provide the <code>ref</code> names of the reference panel. If <code>ref</code> names are provided as “EUR”, “AFR”, “EAS”, “SAS” ,or “AMR”, then the default 1000 genome project reference panels will be used; otherwise <code>ref</code> needs to be provided as a directory to the plink1 format files (.bim, .bed, .fam).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BRIGHT</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority GWAS summary statistics file or marginal genotype-trait inner product file###</span></span>
<span></span>
<span><span class="co"># Read in target GWAS</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="st">"GWAS"</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in target marginal genotype-trait inner product</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="st">"IProd"</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read prior majority GWAS summary statistics file, marginal genotype-trait inner product, or joint coefficient estimates, more than 1 prior majority data can be read in###</span></span>
<span></span>
<span><span class="va">Pind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GWAS"</span>,<span class="st">"IProd"</span>,<span class="st">"Coef"</span><span class="op">)</span></span>
<span><span class="va">Pss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_GWAS1.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss1</span><span class="op">)</span></span>
<span><span class="va">Pss2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_IProd2.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss2</span><span class="op">)</span></span>
<span><span class="va">Pss3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef3.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss3</span><span class="op">)</span></span>
<span><span class="va">Pss</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="va">Pss1</span>,<span class="st">"2"</span><span class="op">=</span><span class="va">Pss2</span>,<span class="st">"3"</span><span class="op">=</span><span class="va">Pss3</span><span class="op">)</span> <span class="co"># The order of list Pss need to be matched with Pind</span></span>
<span></span>
<span><span class="co">### Specify the PLINK file stub of the reference panel or "EUR", "AFR", "EAS", "SAS" ,or "AMR" ###</span></span>
<span><span class="va">ref.bfile</span> <span class="op">&lt;-</span> <span class="st">"refpanel"</span></span>
<span></span>
<span><span class="co">### Read LD region file, only required if ref.bfile is provided as PLINK1 format ###</span></span>
<span><span class="va">LDblocks</span> <span class="op">&lt;-</span> <span class="st">"AFR.hg19"</span> <span class="co"># This will use LD regions as defined in Berisa and Pickrell (2015) for the African population and the hg19 genome build.</span></span>
<span><span class="co"># Other alternatives available. Type ?BRIGHTs for more details. </span></span></code></pre></div>
<p>Reference: <a href="https://academic.oup.com/bioinformatics/article/32/2/283/1743626/Approximately-independent-linkage-disequilibrium" class="external-link">Berisa and Pickrell (2015)</a></p>
<p>Then, a preprocessing step is required to remove the SNPs that are not in the reference panel from all data, convert target data into marginal SNPs-trait inner product, convert prior data into joint coefficient estimates, and match the effect alleles between the reference panel and data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">PreprocessS</span><span class="op">(</span>Tss <span class="op">=</span> <span class="va">Tss</span>, Tind <span class="op">=</span> <span class="va">Tind</span>, Pss <span class="op">=</span> <span class="va">Pss</span>, Pind <span class="op">=</span> <span class="va">Pind</span>, ref.bfile<span class="op">=</span><span class="va">ref.bfile</span>, LDblocks<span class="op">=</span><span class="va">LDblocks</span><span class="op">)</span></span></code></pre></div>
<p>Running BRIGHTs using standard pipeline with LASSO penalty on quantitative traits:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/BRIGHTs.html">BRIGHTs</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>, type.trait<span class="op">=</span><span class="st">"quantitative"</span>, penalty<span class="op">=</span><span class="st">"LASSO"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="brights-with-binary-traits">BRIGHTs with binary traits<a class="anchor" aria-label="anchor" href="#brights-with-binary-traits"></a>
</h2>
<p>For binary traits, in addition to the GWAS summary statistics or marginal genotype-trait inner product, <span class="math inline">$\frac{\boldsymbol X^\top\boldsymbol y}{n}$</span>, BRIGHTs requires an estimate based on logistic LASSO regression, <span class="math inline"><strong>b̂</strong></span>, and the marginal genotype-predicted traits inner product, <span class="math inline">$\frac{\boldsymbol X^\top expit(\boldsymbol X \boldsymbol{\hat b})}{n}$</span>, from the target minority population. From the prior majority populations coefficients estimated from joint models (e.g. logistic LASSO regression) is required for model fitting. We note that more than 1 prior majority data can be incorporated in the BRIGHTs model.</p>
<p>First we read the minority summary statistics and majority summary statistics into R, and provide the <code>ref</code> names of the reference panel. If <code>ref</code> names are provided as “EUR”, “AFR”, “EAS”, “SAS” ,or “AMR”, then the default 1000 genome project reference panels will be used; otherwise <code>ref</code> needs to be provided as a directory to the plink1 format files (.bim, .bed, .fam).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BRIGHT</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority GWAS summary statistics file or marginal genotype-trait inner product file###</span></span>
<span></span>
<span><span class="co"># Read in target GWAS</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GWAS"</span>,<span class="st">"LASSO"</span>,<span class="st">"IProdPred"</span><span class="op">)</span></span>
<span><span class="va">Tss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in target marginal genotype-trait inner product</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IProd"</span>,<span class="st">"LASSO"</span>,<span class="st">"IProdPred"</span><span class="op">)</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority LASSO estimates file and marginal genotype-predicted outcome inner product file###</span></span>
<span><span class="va">bhat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_LASSO.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bhat</span><span class="op">)</span></span>
<span><span class="va">rhat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProdPred.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bhat</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span> <span class="op">=</span> <span class="va">Tss1</span>, <span class="st">"2"</span> <span class="op">=</span> <span class="va">bhat</span>, <span class="st">"3"</span> <span class="op">=</span> <span class="va">rhat</span><span class="op">)</span> <span class="co"># The order of list Tss need to be matched with Tind</span></span>
<span></span>
<span><span class="co">### Read prior majority GWAS summary statistics file, marginal genotype-trait inner product, or joint coefficient estimates, more than 1 prior majority data can be read in###</span></span>
<span></span>
<span><span class="va">Pind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Coef"</span>, <span class="st">"Coef"</span>, <span class="st">"Coef"</span><span class="op">)</span></span>
<span><span class="va">Pss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef1.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss1</span><span class="op">)</span></span>
<span><span class="va">Pss2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef2.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss2</span><span class="op">)</span></span>
<span><span class="va">Pss3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef3.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss3</span><span class="op">)</span></span>
<span><span class="va">Pss</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="va">Pss1</span>,<span class="st">"2"</span><span class="op">=</span><span class="va">Pss2</span>,<span class="st">"3"</span><span class="op">=</span><span class="va">Pss3</span><span class="op">)</span> <span class="co"># The order of list Pss need to be matched with Pind</span></span>
<span></span>
<span><span class="co">### Specify the PLINK file stub of the reference panel or "EUR", "AFR", "EAS", "SAS" ,or "AMR" ###</span></span>
<span><span class="va">ref.bfile</span> <span class="op">&lt;-</span> <span class="st">"refpanel"</span></span>
<span></span>
<span><span class="co">### Read LD region file, only required if ref.bfile is provided as PLINK1 format ###</span></span>
<span><span class="va">LDblocks</span> <span class="op">&lt;-</span> <span class="st">"AFR.hg19"</span> <span class="co"># This will use LD regions as defined in Berisa and Pickrell (2015) for the African population and the hg19 genome build.</span></span>
<span><span class="co"># Other alternatives available. Type ?BRIGHTs for more details. </span></span></code></pre></div>
<p>Then, a preprocessing step is required to remove the SNPs that are not in the reference panel from all data, convert target data into marginal SNPs-trait inner product, convert prior data into joint coefficient estimates, and match the effect alleles between the reference panel and data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">PreprocessS</span><span class="op">(</span>Tss <span class="op">=</span> <span class="va">Tss</span>, Tind <span class="op">=</span> <span class="va">Tind</span>, Pss <span class="op">=</span> <span class="va">Pss</span>, Pind <span class="op">=</span> <span class="va">Pind</span>, ref.bfile<span class="op">=</span><span class="va">ref.bfile</span>, LDblocks<span class="op">=</span><span class="va">LDblocks</span><span class="op">)</span></span></code></pre></div>
<p>Running BRIGHTs using standard pipeline with LASSO penalty on quantitative traits:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/BRIGHTs.html">BRIGHTs</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>, type.trait<span class="op">=</span><span class="st">"binary"</span>, penalty<span class="op">=</span><span class="st">"LASSO"</span><span class="op">)</span></span></code></pre></div>
<p>This procedure requires additional and quite stringent summary statistics from both target and prior data, in genetics studies its quite common to treat binary outcome as continuous and perform continuous models on the data; therefore, in the case where the above additonal summary statistics are not available, the BRIGHTS with quantitative traits procedure can also be used to analyze the binary data.</p>
</div>
</div>
<div class="section level1">
<h1 id="brighti-tutorial">BRIGHTi tutorial<a class="anchor" aria-label="anchor" href="#brighti-tutorial"></a>
</h1>
<p>BRIGHTi group of methods utilize individual-level data from target minority populations and a wide variety of summary-level data from prior majority population to carry out transfer-learning. Summary statistics are expected to be loaded into memory as a data.frame/data.table for the prior majority population.</p>
<p>Below we discuss the required data and implementation tutorials.</p>
<p>BRIGHTi requires the individual-level genotype and phenotype from the target minority population, while from the prior majority populations either GWAS summary statistics, marginal genotype-trait inner product, or coefficients estimated from joint models (e.g. PRS or LASSO regression) can be used for model fitting. We note that more than 1 prior majority data can be incorporated in the BRIGHTi model.</p>
<p>First we read the minority genotype data from plink1 files and phenotype data from text files and majority summary statistics into R.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BRIGHT</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority GWAS summary statistics file or marginal genotype-trait inner product file###</span></span>
<span></span>
<span><span class="co"># Read in target individual-level data</span></span>
<span><span class="va">Tgeno</span> <span class="op">&lt;-</span> <span class="st">"/path/to/plink"</span></span>
<span><span class="va">Tpheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_phenotype.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tpheno</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">### Read prior majority GWAS summary statistics file, marginal genotype-trait inner product, or joint coefficient estimates, more than 1 prior majority data can be read in###</span></span>
<span></span>
<span><span class="va">Pind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GWAS"</span>,<span class="st">"IProd"</span>,<span class="st">"Coef"</span><span class="op">)</span></span>
<span><span class="va">Pss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_GWAS1.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss1</span><span class="op">)</span></span>
<span><span class="va">Pss2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_IProd2.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss2</span><span class="op">)</span></span>
<span><span class="va">Pss3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef3.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss3</span><span class="op">)</span></span>
<span><span class="va">Pss</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="va">Pss1</span>,<span class="st">"2"</span><span class="op">=</span><span class="va">Pss2</span>,<span class="st">"3"</span><span class="op">=</span><span class="va">Pss3</span><span class="op">)</span> <span class="co"># The order of list Pss need to be matched with Pind</span></span></code></pre></div>
<p>Then, a preprocessing step is required to remove the SNPs that are not in the target minority genotype files from prior majority data, convert prior data into joint coefficient estimates, and match the effect alleles between the minority genotype and prior data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">PreprocessI</span><span class="op">(</span>Tpheno <span class="op">=</span> <span class="va">Tpheno</span>, Tgeno <span class="op">=</span> <span class="va">Tgeno</span>, Pss <span class="op">=</span> <span class="va">Pss</span>, Pind <span class="op">=</span> <span class="va">Pind</span><span class="op">)</span></span></code></pre></div>
<p>Running BRIGHTi using standard pipeline with LASSO penalty on different types of traits including “quantitative”, “binary”, “count”:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/BRIGHTi.html">BRIGHTi</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>, type.trait<span class="op">=</span><span class="st">"quantitative"</span>, penalty<span class="op">=</span><span class="st">"LASSO"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="model-validation">Model validation<a class="anchor" aria-label="anchor" href="#model-validation"></a>
</h1>
<div class="section level2">
<h2 id="model-validation-with-individual-level-test-data">Model validation with individual-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-individual-level-test-data"></a>
</h2>
<p>When individual-level test data is available, <code>BRIGHT</code> package provide automated validation functions and generates evaluation plots:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in target individual-level data</span></span>
<span><span class="va">Testgeno</span> <span class="op">&lt;-</span> <span class="st">"/path/to/test/plink"</span></span>
<span><span class="va">Testpheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Test_phenotype.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testpheno</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Ind</span><span class="op">(</span><span class="va">out</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="model-validation-with-summary-level-test-data">Model validation with summary-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-summary-level-test-data"></a>
</h2>
<p>When summary test data is available, <code>BRIGHT</code> package provide automated validation functions for parameter fine-tunning:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in test GWAS</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"GWAS"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in test marginal genotype-trait inner product</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"IProd"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Sum</span><span class="op">(</span><span class="va">Testss</span>, <span class="va">Testind</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
<p>We note that summary level test data is only supported for quantitative traits</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 2)</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing BRIGHT</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Qinmengge Li <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Qinmengge Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
