<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BRIGHT • BRIGHT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BRIGHT">
<meta property="og:description" content="BRIGHT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BRIGHT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/BRIGHT.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/BRIGHTi.html">BRIGHTi tutorial</a>
    </li>
    <li>
      <a href="../articles/BRIGHTs.html">BRIGHTs tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BRIGHT</h1>
                        <h4 data-toc-skip class="author">Qinmengge Li</h4>
            
            <h4 data-toc-skip class="date">5/1/2023</h4>
      
      
      <div class="hidden name"><code>BRIGHT.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Polygenic risk scores (PRS) are commonly used tools for genetic risk prediction . However, these scores have been primarily based on European ancestry or East Asian ancestry cohort, with limited prediction power for minority cohorts (e.g. African ancestry), known as the lack of model transferability . In contrast, PRS constructed directly from minority ethnic data have long suffered small sample size and “the curse of dimensionality” . Hence, it would be desirable to design a procedure to transfer knowledge from the well-studied prior populations to the under-represented target population and optimally balance the information between them to account for the heterogeneity. A unique challenge here is that, due to the DataSHIELD constraint , individual data are not accessible across studies. Instead, only summary statistics, such as the SNP-trait correlation, are available, which precludes the application of existing transfer learning procedures. In response, we proposed a privacy-preserving technique, BRegman divergence-based Integrational Genetic Hazard Trans-ethnic (BRIGHT), using published summary statistics from different ethnic populations to improve the recalibration, discrimination, and prediction accuracy on the target minority cohort .</p>
<p>BRIGHT estimation procedure utilize Bregman-divergence to measure the difference between the populations and optimally balance the information from them. The method enjoys the following properties : (1) it controls the relative weight of the external information, identifying the most compatible ones and diminishing the weights of less relevant ones; (2) it is robust to model misspecification and incorporates various types of external risk scores, including risk scores derived from different statistical models or built upon different outcomes; (3) it accommodates that the covariate space may be different between data sources; (4) it requires less sensitive summary statistics to be obtained from both target and prior data sources, with no use of any sensitive individual-level data; (5) it achieves fine-mapping simultaneously in the process of PRS construction; (6) its objective function retains similar forms as the penalized regression and guarantees the computation efficiency; and (7) it incorporates pLASSO  and LASSOsum  as special cases.</p>
<p>The software package  implements the above BRIGHT estimation procedure. Within the package, two versions of BRIGHT models are implemented: (1) BRIGHTi, which requires the individual-level target minority data and summary-level prior data; (2) BRIGHTs, which only requires the summary-level information from both target and prior data, therefore preserves patient privacy. Both BRIGHTi and BRIGHTs methods are efficiently implemented with general penalty structures including, group LASSO, group SCAD, and group MCP, which enable the group genetic marker fine mapping and incorporate LASSO, SCAD, and MCP penalties as special cases.</p>
<p>This vignette describes the usage of BRIGHT with summary level data and quantitative traits in R. There are additional vignettes that should be useful:</p>
<ul>
<li>“BRIGHTs” describes the usage of BRIGHT with summary level target data and both binary and quantitative traits.</li>
<li>“BRIGHTi” describes the usage of BRIGHT with individual level target data and binary, count, and quantitative traits.</li>
</ul>
<p>To integrate summary statistics without using individual-level data, while adjusting for the potential disparities between prior majority and target minority populations, we propose the following penalized objective function, <span class="math display">\[\begin{align}
Q_{BRIGHTs}(\boldsymbol\beta) =&amp; B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta) + \eta B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\check\beta}||\boldsymbol\beta) + p_{\lambda}(\boldsymbol\beta),\label{eq: BRIGHT}
\end{align}\]</span> where <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> and <span class="math inline">\(\boldsymbol{\check\beta}\)</span> are the estimated coefficients obtained from published studies based on minority and majority populations; <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\)</span> is the regularized block-structured LD estimation  from publicly available minority genotypes (e.g. 1000 genome project). <span class="math inline">\(B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)=(\boldsymbol{\hat\beta}-\boldsymbol{\beta})^\top\widetilde{\boldsymbol\Sigma}(\boldsymbol{\hat\beta}-\boldsymbol{\beta})/n\)</span> is the Bregman-divergence between <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> and <span class="math inline">\(\boldsymbol{\beta}\)</span>, serving as an approximation of the negative log-likelihood from the minority data, which is not directly available due to DataSHIELD constraints; <span class="math inline">\(B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\check\beta}||\boldsymbol\beta)=(\boldsymbol{\check\beta}-\boldsymbol{\beta})^\top\widetilde{\boldsymbol\Sigma}(\boldsymbol{\check\beta}-\boldsymbol{\beta})/n\)</span> is the Bregman-divergence between <span class="math inline">\(\boldsymbol{\check\beta}\)</span> and <span class="math inline">\(\boldsymbol{\beta}\)</span>, and is an approximation of the Bregman-divergence in (). The latter approximation is achieved by <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\approx\widehat{\boldsymbol\Sigma}\)</span>, which is widely deemed to be true in genetics studies .</p>
<p>For quantitative traits, <span class="math inline">\(\boldsymbol{\hat\beta}=\widetilde{\boldsymbol\Sigma}^{-1}\boldsymbol r\)</span> ensures <span class="math inline">\(B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)\)</span> to be a valid approximation of the OLS loss through <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\approx\widehat{\boldsymbol\Sigma}\)</span> , where <span class="math inline">\(\boldsymbol r=\frac{\boldsymbol y^\top\boldsymbol X}{n}\)</span> is the marginal SNPs-trait inner product and a standardized version, <span class="math inline">\(\boldsymbol r^*=\frac{\boldsymbol y^{*\top}\boldsymbol X^*}{n}\)</span>, can be recovered from GWAS summary statistics with <span class="math inline">\(\boldsymbol X^*\)</span> and <span class="math inline">\(\boldsymbol y^*\)</span> being the standardized genotype and phenotype . We note, in this scenario, <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> is only presented for illustration purposes, the actual implementation does not require the inversion or invertibility of <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\)</span>; furthermore, the BRIGHTs method will reduce to LASSOsum  when <span class="math inline">\(\eta=0\)</span> and <span class="math inline">\(p_{\lambda}(\boldsymbol\beta)\)</span> is chosen as LASSO penalty. The pipeline of the BRIGHTs estimation procedure with quantitative traits and target minority GWAS summary statistics is presented in Fig.A, and this pipeline is used for the case study in Section .</p>
<p>We note that BRIGHT estimation procedure can incorporate a wide variety of summary-level information from the prior majority population for the construction of <span class="math inline">\(\boldsymbol{\check\beta}\)</span>: (1) when PRS coefficients from the prior majority population are available, they can be directly employed as <span class="math inline">\(\boldsymbol{\check\beta}\)</span>; (2) when only GWAS summary statistics from prior majority population are avialble, <span class="math inline">\(\boldsymbol{\check\beta}\)</span> can be estimated through LASSOsum PRS method; (3) when individual-level data from the prior majority population can be accessed by some collaborators, <span class="math inline">\(\boldsymbol{\check\beta}\)</span> can be estimated from LASSO regression and only the LASSO coefficient estimates need to be shared. In addition, <span class="math inline">\(\boldsymbol{\check\beta}\)</span> is only required to contain a subset of markers in the target minority genotype, <span class="math inline">\(\boldsymbol X\)</span>, which makes the BRIGHTi estimation procedure more flexible than the current trans-ethnic PRS approaches.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>BRIGHT</code> requires the following R packages: <code>Rcpp</code>, <code>Matrix</code>, <code>mvtnorm</code>, <code>BEDMatrix</code>, <code>grpreg</code>. Install them by:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Rcpp"</span>, <span class="st">"Matrix"</span>, <span class="st">"mvtnorm"</span>, <span class="st">"BEDMatrix"</span>, <span class="st">"grpreg"</span><span class="op">)</span>, dependencies<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>For Windows and Mac users, it would be easiest to download the following binaries (<a href="To%20be%20determined">Windows</a>, <a href="To%20be%20determined">Mac</a>) and install using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"/path/to/downloaded_binary_file"</span>, repos<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p>If you are on Linux or you would like to compile from source, you can download the source codes <a href="To%20be%20determined">BRIGHT_0.0.1.tar.gz</a>. Mac users should refer to <a href="https://cran.r-project.org/bin/macosx/tools/" class="external-link">this page</a> for the various dependencies required. Install then via:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"/path/to/downloaded_source.tar.gz"</span>, repos<span class="op">=</span><span class="cn">NULL</span>, type<span class="op">=</span><span class="st">"source"</span><span class="op">)</span></span></code></pre></div>
<p>If you have <code>devtools</code>, you can also type:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install_github</span><span class="op">(</span><span class="st">"To be determined"</span><span class="op">)</span></span></code></pre></div>
<p>or</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install_github</span><span class="op">(</span><span class="st">"To be determined"</span><span class="op">)</span></span></code></pre></div>
<p>for the latest development version. Or you can clone the latest development version here and install yourself using <code>devtools</code>.</p>
</div>
<div class="section level2">
<h2 id="warning">Warning!<a class="anchor" aria-label="anchor" href="#warning"></a>
</h2>
<p>Most functions in <code>BRIGHT</code> impute missing genotypes in PLINK bfiles with a homozygous A2 genotype, which is the same as using the <code>--fill-missing-a2</code> option in PLINK. It is the user’s responsibility to filter out individuals and SNPs with too many missing genotypes beforehand.</p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>BRIGHTs group of methods utilize a wide variety of summary-level data from different populations to carry out transfer-learning. We accounted for Linkage Disequilibrium (LD) via a reference panel (1000 genome project as default). The reference panel is assumed to be in PLINK 1 <a href="https://www.cog-genomics.org/plink/1.9/input#bed" class="external-link">format</a>. Summary statistics are expected to be loaded into memory as a data.frame/data.table.</p>
<p>Below we discuss the required data and implementation tutorials separately for quantitative traits and binary traits.</p>
<div class="section level3">
<h3 id="brights-with-quantitative-traits">BRIGHTs with quantitative traits<a class="anchor" aria-label="anchor" href="#brights-with-quantitative-traits"></a>
</h3>
<p>For quantitative traits, BRIGHTs requires the GWAS summary statistics or marginal genotype-trait inner product, <span class="math inline">\(\frac{\boldsymbol X^\top\boldsymbol y}{n}\)</span>, from the target minority population, while from the prior majority populations either GWAS summary statistics, marginal genotype-trait inner product, or coefficients estimated from joint models (e.g. PRS or LASSO regression) can be used for model fitting. We note that more than 1 prior majority data can be incorporated in the BRIGHTs model.</p>
<p>First we read the minority summary statistics and majority summary statistics into R, and provide the <code>ref</code> names of the reference panel. If <code>ref</code> names are provided as “EUR”, “AFR”, “EAS”, “SAS” ,or “AMR”, then the default 1000 genome project reference panels will be used; otherwise <code>ref</code> needs to be provided as a directory to the plink1 format files (.bim, .bed, .fam).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BRIGHT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BRIGHT'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     norm</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Read target minority GWAS summary statistics file or marginal genotype-trait inner product file###</span></span>
<span></span>
<span><span class="co"># Read in target GWAS</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="st">"GWAS"</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"~/Desktop/research/Kevin He/PRS/Rewrite_Lassosum/example/SAS1.GWAS"</span>,header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   CHR     BP A1 A2 Sign    N      P</span></span>
<span><span class="co">## 1   1 753405  A  C    1 1760 0.3624</span></span>
<span><span class="co">## 2   1 754192  G  A    1 1760 0.3065</span></span>
<span><span class="co">## 3   1 757640  A  G    1 1760 0.4178</span></span>
<span><span class="co">## 4   1 758626  T  C    1 1760 0.5019</span></span>
<span><span class="co">## 5   1 769223  G  C   -1 1760 0.6513</span></span>
<span><span class="co">## 6   1 777122  T  A    1 1760 0.5741</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Read prior majority GWAS summary statistics file, marginal genotype-trait inner product, or joint coefficient estimates, more than 1 prior majority data can be read in###</span></span>
<span></span>
<span><span class="va">Pind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Coef"</span><span class="op">)</span></span>
<span><span class="va">Pss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"~/Desktop/research/Kevin He/PRS/Rewrite_Lassosum/example/EUR.Coef"</span>,header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   CHR     BP A1 A2        Coef</span></span>
<span><span class="co">## 1   1 753405  A  C 0.000000000</span></span>
<span><span class="co">## 2   1 754192  G  A 0.000000000</span></span>
<span><span class="co">## 3   1 757640  A  G 0.000000000</span></span>
<span><span class="co">## 4   1 758626  T  C 0.000000000</span></span>
<span><span class="co">## 5   1 769223  G  C 0.000000000</span></span>
<span><span class="co">## 6   1 777122  T  A 0.001384477</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Specify the PLINK file stub of the reference panel or "EUR", "AFR", "EAS", "SAS" ,or "AMR" ###</span></span>
<span><span class="va">Tref</span> <span class="op">&lt;-</span> <span class="st">"~/Desktop/research/Kevin He/PRS/Rewrite_Lassosum/example/SAS1"</span></span>
<span><span class="va">Pref</span> <span class="op">&lt;-</span> <span class="st">"~/Desktop/research/Kevin He/PRS/Rewrite_Lassosum/example/EUR"</span></span>
<span></span>
<span><span class="co">### Read LD region file, only required if ref.bfile is provided as PLINK1 format ###</span></span>
<span><span class="va">LDblocks</span> <span class="op">&lt;-</span> <span class="st">"SAS"</span> <span class="co"># This will use LD regions as defined in Berisa and Pickrell (2015) for the South Asian population and the hg19 genome build.</span></span>
<span><span class="co"># Other alternatives available. Type ?BRIGHTs for more details. </span></span></code></pre></div>
<p>Reference: <a href="https://academic.oup.com/bioinformatics/article/32/2/283/1743626/Approximately-independent-linkage-disequilibrium" class="external-link">Berisa and Pickrell (2015)</a></p>
<p>Then, a preprocessing step is required to remove the SNPs that are not in the reference panel from all data, convert target data into marginal SNPs-trait inner product, convert prior data into joint coefficient estimates, and match the effect alleles between the reference panel and data.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PreprocessS.html">PreprocessS</a></span><span class="op">(</span><span class="va">Tss</span>, <span class="va">Tind</span>, <span class="va">Tref</span>, <span class="va">Pss</span>, <span class="va">Pind</span>, <span class="va">Pref</span>, <span class="va">LDblocks</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using user specified LD</span></span>
<span><span class="co">## 1 source of prior information identified</span></span>
<span><span class="co">## 5000 SNPs passed filtering for target data</span></span>
<span><span class="co">## 5000 SNPs passed filtering for prior data 1</span></span></code></pre>
<p>Running BRIGHTs using standard pipeline with LASSO penalty on quantitative traits:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BRIGHTs.html">BRIGHTs</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>, type.trait<span class="op">=</span><span class="st">"quantitative"</span>, penalty<span class="op">=</span><span class="st">"LASSO"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="model-validation-and-visualization">Model validation and visualization<a class="anchor" aria-label="anchor" href="#model-validation-and-visualization"></a>
</h3>
<div class="section level4">
<h4 id="model-validation-with-individual-level-test-data">Model validation with individual-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-individual-level-test-data"></a>
</h4>
<p>When individual-level test data is available, <code>BRIGHT</code> package provide automated validation functions and generates evaluation plots:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in target individual-level data</span></span>
<span><span class="va">Testgeno</span> <span class="op">&lt;-</span> <span class="st">"~/Desktop/research/Kevin He/PRS/Rewrite_Lassosum/example/SAS2"</span></span>
<span><span class="va">Testpheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"~/Desktop/research/Kevin He/PRS/Rewrite_Lassosum/example/SAS2_phe.fam"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testpheno</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   V1     V2         V3</span></span>
<span><span class="co">## 1  0 711676   9.415242</span></span>
<span><span class="co">## 2  0 711677  20.729158</span></span>
<span><span class="co">## 3  0 711679   8.834436</span></span>
<span><span class="co">## 4  0 711680   8.004271</span></span>
<span><span class="co">## 5  0 711681 -11.331491</span></span>
<span><span class="co">## 6  0 711682  -2.485393</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Ind</span><span class="op">(</span><span class="va">out</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot MSPE and Cor plot</span></span>
<span><span class="fu">MSE_Cor.plot</span><span class="op">(</span><span class="va">Val</span><span class="op">)</span></span></code></pre></div>
<p><img src="BRIGHT_files/figure-html/unnamed-chunk-6-1.png" width="960"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot MSPE and Cor plot</span></span>
<span><span class="fu">Density.plot</span><span class="op">(</span><span class="va">Val</span>,Pct<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span></span></code></pre></div>
<p><img src="BRIGHT_files/figure-html/unnamed-chunk-7-1.png" width="960"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot MSPE and Cor plot</span></span>
<span><span class="fu">ROC.plot</span><span class="op">(</span><span class="va">Val</span>,Pct<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span></span></code></pre></div>
<p><img src="BRIGHT_files/figure-html/unnamed-chunk-8-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="model-validation-with-summary-level-test-data">Model validation with summary-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-summary-level-test-data"></a>
</h4>
<p>When summary test data is available, <code>BRIGHT</code> package provide automated validation functions for parameter fine-tunning:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in test GWAS</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"GWAS"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu">fread</span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in test marginal genotype-trait inner product</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"IProd"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu">fread</span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Sum</span><span class="op">(</span><span class="va">Testss</span>, <span class="va">Testind</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
<p>We note that summary level test data is only supported for quantitative traits</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Qinmengge Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
