<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>BRIGHTi tutorial • BRIGHT</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="BRIGHTi tutorial">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BRIGHT</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/BRIGHT.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/BRIGHTi.html">BRIGHTi tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/BRIGHTs.html">BRIGHTs tutorial</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/UM-KevinHe/BRIGHT/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>BRIGHTi tutorial</h1>
                        <h4 data-toc-skip class="author">Qinmengge
Li</h4>
            
            <h4 data-toc-skip class="date">5/1/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UM-KevinHe/BRIGHT/blob/HEAD/vignettes/BRIGHTi.Rmd" class="external-link"><code>vignettes/BRIGHTi.Rmd</code></a></small>
      <div class="d-none name"><code>BRIGHTi.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Let genotype,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐗</mi><annotation encoding="application/x-tex">\boldsymbol X</annotation></semantics></math>,
and phenotype,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐲</mi><annotation encoding="application/x-tex">\boldsymbol y</annotation></semantics></math>,
from the target minority population, and coefficient vector,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>𝛃</mi><mo accent="true">̌</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\check\beta}</annotation></semantics></math>,
obtained from the published PRS model for the prior majority population
are available. To integrate data while adjusting for the potential
disparities among different populations, we combine the log-likelihood
loss function from the target minority data and a
Bregman-divergence-based metric to construct a penalized objective
function,
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Q</mi><mrow><mi>B</mi><mi>R</mi><mi>I</mi><mi>G</mi><mi>H</mi><mi>T</mi><mi>i</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>−</mi><mo>ℓ</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>η</mi><msub><mi>B</mi><mover><mi>𝚺</mi><mo accent="true">̂</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>𝛃</mi><mo accent="true">̌</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><msub><mi>p</mi><mi>λ</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{align}
Q_{BRIGHTi}(\boldsymbol\beta) = -\ell(\boldsymbol\beta)+\eta B_{\widehat{\boldsymbol\Sigma}}(\boldsymbol{\check\beta}||\boldsymbol\beta)+p_{\lambda}(\boldsymbol\beta).\label{eq: BRIGHTi}
\end{align}</annotation></semantics></math></p>
<p>For quantitative trait,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>ℓ</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>−</mi><msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐲</mi><mo>−</mo><mi>𝐗</mi><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>⊤</mi></msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐲</mi><mo>−</mo><mi>𝐗</mi><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>/</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">\ell(\boldsymbol\beta)=-(\boldsymbol y-\boldsymbol X\boldsymbol{\beta})^\top(\boldsymbol y-\boldsymbol X\boldsymbol{\beta})/n</annotation></semantics></math>,
is the OLS loss function measuring the discrepancy between the working
model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛃</mi><annotation encoding="application/x-tex">\boldsymbol\beta</annotation></semantics></math>,
and the target minority data;
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>B</mi><mover><mi>𝚺</mi><mo accent="true">̂</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>𝛃</mi><mo accent="true">̌</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>𝛃</mi><mo accent="true">̌</mo></mover><mo>−</mo><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>⊤</mi></msup><mover><mi>𝚺</mi><mo accent="true">̂</mo></mover><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>𝛃</mi><mo accent="true">̌</mo></mover><mo>−</mo><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">B_{\widehat{\boldsymbol\Sigma}}(\boldsymbol{\check\beta}||\boldsymbol\beta)=(\boldsymbol{\check\beta}-\boldsymbol{\beta})^\top\widehat{\boldsymbol\Sigma}(\boldsymbol{\check\beta}-\boldsymbol{\beta})</annotation></semantics></math>,
is a special case of the Bregman-divergence (Mahalanobis distance)
measuring the discrepancy between the working model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛃</mi><annotation encoding="application/x-tex">\boldsymbol\beta</annotation></semantics></math>,
and the prior information,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>𝛃</mi><mo accent="true">̌</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\check\beta}</annotation></semantics></math>,
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>𝚺</mi><mo accent="true">̂</mo></mover><mo>=</mo><msup><mi>𝐗</mi><mi>⊤</mi></msup><mi>𝐗</mi><mi>/</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">\widehat{\boldsymbol\Sigma}=\boldsymbol X^\top \boldsymbol X/n</annotation></semantics></math>;
we note the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>l</mi><mn>2</mn></msub><annotation encoding="application/x-tex">l_2</annotation></semantics></math>
penalty is a special case of the proposed Bregman-divergence.</p>
<p>For qualitative trait,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>ℓ</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><msub><mi>y</mi><mi>i</mi></msub><msubsup><mi>𝐗</mi><mi>i</mi><mi>⊤</mi></msubsup><mi>𝛃</mi><mo>+</mo><mi>b</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>𝐗</mi><mi>i</mi><mi>⊤</mi></msubsup><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mrow><annotation encoding="application/x-tex">\ell(\boldsymbol\beta)=\sum_{i=1}^n {y_i\boldsymbol{X}_i^\top \boldsymbol\beta+b(\boldsymbol{X}_i^\top \boldsymbol\beta)}</annotation></semantics></math>,
is the log-likelihood function measuring the discrepancy between the
working model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛃</mi><annotation encoding="application/x-tex">\boldsymbol\beta</annotation></semantics></math>,
and the target minority data;
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>B</mi><mover><mi>𝚺</mi><mo accent="true">̂</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>𝛃</mi><mo accent="true">̌</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><msubsup><mi>𝐗</mi><mi>i</mi><mi>⊤</mi></msubsup><mover><mi>𝛃</mi><mo accent="true">̌</mo></mover><msubsup><mi>𝐗</mi><mi>i</mi><mi>⊤</mi></msubsup><mi>𝛃</mi><mo>+</mo><mi>b</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>𝐗</mi><mi>i</mi><mi>⊤</mi></msubsup><mi>𝛃</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mrow><annotation encoding="application/x-tex">B_{\widehat{\boldsymbol\Sigma}}(\boldsymbol{\check\beta}||\boldsymbol\beta)=\sum_{i=1}^n {\boldsymbol{X}_i^\top \boldsymbol{\check\beta}\boldsymbol{X}_i^\top \boldsymbol\beta+b(\boldsymbol{X}_i^\top \boldsymbol\beta)}</annotation></semantics></math>,
is another special case of the Bregman-divergence (Kullback-Leibler
distance) measuring the discrepancy between the working model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝛃</mi><annotation encoding="application/x-tex">\boldsymbol\beta</annotation></semantics></math>,
and the prior information,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>𝛃</mi><mo accent="true">̌</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\check\beta}</annotation></semantics></math>.</p>
<p>In addition,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>η</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
is the weight balancing the importance of the minority data over the
prior information to account for the population heterogeneity; when
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>η</mi><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\eta=0</annotation></semantics></math>,
BRIGHTi will reduce to the penalized regression on the target data; when
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>η</mi><mo>→</mo><mi>∞</mi></mrow><annotation encoding="application/x-tex">\eta\rightarrow\infty</annotation></semantics></math>,
BRIGHTi will reduce to the penalized prior information.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mi>λ</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>β</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">p_{\lambda}(\beta)</annotation></semantics></math>
is a general penalty including (group) LASSO, (group) SCAD, and (group)
MCP to achieve sparsity and fine mapping in the genotype effect size
estimation. In the case of quantitative trait, the BRIGHTi estimation
procedure includes pLASSO as a special case when outcome
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐲</mi><annotation encoding="application/x-tex">\boldsymbol y</annotation></semantics></math>
are assumed to follow Gaussian distribution, which the BRIGHTi method
does not require.</p>
<p>BRIGHTi group of methods utilize individual-level data from target
minority populations and a wide variety of summary-level data from prior
majority population to carry out transfer-learning. Summary statistics
are expected to be loaded into memory as a data.frame/data.table for the
prior majority population.</p>
<p>Below we discuss the required data and implementation tutorials.</p>
<p>BRIGHTi requires the individual-level genotype and phenotype from the
target minority population, while from the prior majority populations
either GWAS summary statistics, marginal genotype-trait inner product,
or coefficients estimated from joint models (e.g. PRS or LASSO
regression) can be used for model fitting. We note that more than 1
prior majority data can be incorporated in the BRIGHTi model.</p>
<p>First we read the minority genotype data from plink1 files and
phenotype data from text files and majority summary statistics into
R.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/UM-KevinHe/BRIGHT" class="external-link">BRIGHT</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority GWAS summary statistics file or marginal genotype-trait inner product file###</span></span>
<span></span>
<span><span class="co"># Read in target individual-level data</span></span>
<span><span class="va">Tgeno</span> <span class="op">&lt;-</span> <span class="st">"/path/to/plink"</span></span>
<span><span class="va">Tpheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_phenotype.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tpheno</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">### Read prior majority GWAS summary statistics file, marginal genotype-trait inner product, or joint coefficient estimates, more than 1 prior majority data can be read in###</span></span>
<span></span>
<span><span class="va">Pind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GWAS"</span>,<span class="st">"IProd"</span>,<span class="st">"Coef"</span><span class="op">)</span></span>
<span><span class="va">Pss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_GWAS1.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss1</span><span class="op">)</span></span>
<span><span class="va">Pss2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_IProd2.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss2</span><span class="op">)</span></span>
<span><span class="va">Pss3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef3.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss3</span><span class="op">)</span></span>
<span><span class="va">Pss</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="va">Pss1</span>,<span class="st">"2"</span><span class="op">=</span><span class="va">Pss2</span>,<span class="st">"3"</span><span class="op">=</span><span class="va">Pss3</span><span class="op">)</span> <span class="co"># The order of list Pss need to be matched with Pind</span></span></code></pre></div>
<p>Then, a preprocessing step is required to remove the SNPs that are
not in the target minority genotype files from prior majority data,
convert prior data into joint coefficient estimates, and match the
effect alleles between the minority genotype and prior data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">PreprocessI</span><span class="op">(</span>Tpheno <span class="op">=</span> <span class="va">Tpheno</span>, Tgeno <span class="op">=</span> <span class="va">Tgeno</span>, Pss <span class="op">=</span> <span class="va">Pss</span>, Pind <span class="op">=</span> <span class="va">Pind</span><span class="op">)</span></span></code></pre></div>
<p>Running BRIGHTi using standard pipeline with LASSO penalty on
different types of traits including “quantitative”, “binary”,
“count”:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">BRIGHTi</span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>, type.trait<span class="op">=</span><span class="st">"quantitative"</span>, penalty<span class="op">=</span><span class="st">"LASSO"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="model-validation">Model validation<a class="anchor" aria-label="anchor" href="#model-validation"></a>
</h2>
<div class="section level3">
<h3 id="model-validation-with-individual-level-test-data">Model validation with individual-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-individual-level-test-data"></a>
</h3>
<p>When individual-level test data is available, <code>BRIGHT</code>
package provide automated validation functions and generates evaluation
plots:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in target individual-level data</span></span>
<span><span class="va">Testgeno</span> <span class="op">&lt;-</span> <span class="st">"/path/to/test/plink"</span></span>
<span><span class="va">Testpheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Test_phenotype.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testpheno</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Ind</span><span class="op">(</span><span class="va">out</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="model-validation-with-summary-level-test-data">Model validation with summary-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-summary-level-test-data"></a>
</h3>
<p>When summary test data is available, <code>BRIGHT</code> package
provide automated validation functions for parameter fine-tunning:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in test GWAS</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"GWAS"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in test marginal genotype-trait inner product</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"IProd"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Sum</span><span class="op">(</span><span class="va">Testss</span>, <span class="va">Testind</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
<p>We note that summary level test data is only supported for
quantitative traits</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qinmengge Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
