<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>BRIGHTs tutorial â€¢ BRIGHT</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="BRIGHTs tutorial">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BRIGHT</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/BRIGHT.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/BRIGHTi.html">BRIGHTi tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/BRIGHTs.html">BRIGHTs tutorial</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/UM-KevinHe/BRIGHT/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>BRIGHTs tutorial</h1>
                        <h4 data-toc-skip class="author">Qinmengge
Li</h4>
            
            <h4 data-toc-skip class="date">5/1/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UM-KevinHe/BRIGHT/blob/HEAD/vignettes/BRIGHTs.Rmd" class="external-link"><code>vignettes/BRIGHTs.Rmd</code></a></small>
      <div class="d-none name"><code>BRIGHTs.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>To integrate summary statistics without using individual-level data,
while adjusting for the potential disparities between prior majority and
target minority populations, we propose the following penalized
objective function,
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Q</mi><mrow><mi>B</mi><mi>R</mi><mi>I</mi><mi>G</mi><mi>H</mi><mi>T</mi><mi>s</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><msub><mi>B</mi><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>Î·</mi><msub><mi>B</mi><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">ÌŒ</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><msub><mi>p</mi><mi>Î»</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{align}
Q_{BRIGHTs}(\boldsymbol\beta) =&amp; B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta) + \eta B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\check\beta}||\boldsymbol\beta) + p_{\lambda}(\boldsymbol\beta),\label{eq: BRIGHT}
\end{align}</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">ÌŒ</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\check\beta}</annotation></semantics></math>
are the estimated coefficients obtained from published studies based on
minority and majority populations;
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover><annotation encoding="application/x-tex">\widetilde{\boldsymbol\Sigma}</annotation></semantics></math>
is the regularized block-structured LD estimation from publicly
available minority genotypes (e.g.Â 1000 genome project).
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>B</mi><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><mo>âˆ’</mo><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>âŠ¤</mi></msup><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><mo>âˆ’</mo><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>/</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)=(\boldsymbol{\hat\beta}-\boldsymbol{\beta})^\top\widetilde{\boldsymbol\Sigma}(\boldsymbol{\hat\beta}-\boldsymbol{\beta})/n</annotation></semantics></math>
is the Bregman-divergence between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>ğ›ƒ</mi><annotation encoding="application/x-tex">\boldsymbol{\beta}</annotation></semantics></math>,
serving as an approximation of the negative log-likelihood from the
minority data, which is not directly available due to DataSHIELD
constraints;
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>B</mi><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">ÌŒ</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">ÌŒ</mo></mover><mo>âˆ’</mo><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>âŠ¤</mi></msup><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">ÌŒ</mo></mover><mo>âˆ’</mo><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>/</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\check\beta}||\boldsymbol\beta)=(\boldsymbol{\check\beta}-\boldsymbol{\beta})^\top\widetilde{\boldsymbol\Sigma}(\boldsymbol{\check\beta}-\boldsymbol{\beta})/n</annotation></semantics></math>
is the Bregman-divergence between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">ÌŒ</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\check\beta}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>ğ›ƒ</mi><annotation encoding="application/x-tex">\boldsymbol{\beta}</annotation></semantics></math>,
and is an approximation of the Bregman-divergence in (). The latter
approximation is achieved by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover><mo>â‰ˆ</mo><mover><mi>ğšº</mi><mo accent="true">Ì‚</mo></mover></mrow><annotation encoding="application/x-tex">\widetilde{\boldsymbol\Sigma}\approx\widehat{\boldsymbol\Sigma}</annotation></semantics></math>,
which is widely deemed to be true in genetics studies .</p>
<p>Although BRIGHTs can incorporate general
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}</annotation></semantics></math>
for data integration, only specific
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}</annotation></semantics></math>
can ensure
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>B</mi><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)</annotation></semantics></math>
to be a valid approximation of the minority data negative
log-likelihood. Below we discuss how to construct
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}</annotation></semantics></math>
from summary-level information in the case of quantitative traits and
binary traits.</p>
<p>For quantitative traits,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><mo>=</mo><msup><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover><mrow><mi>âˆ’</mi><mn>1</mn></mrow></msup><mi>ğ«</mi></mrow><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}=\widetilde{\boldsymbol\Sigma}^{-1}\boldsymbol r</annotation></semantics></math>
ensures
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>B</mi><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)</annotation></semantics></math>
to be a valid approximation of the OLS loss through
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover><mo>â‰ˆ</mo><mover><mi>ğšº</mi><mo accent="true">Ì‚</mo></mover></mrow><annotation encoding="application/x-tex">\widetilde{\boldsymbol\Sigma}\approx\widehat{\boldsymbol\Sigma}</annotation></semantics></math>
, where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ğ«</mi><mo>=</mo><mfrac><mrow><msup><mi>ğ²</mi><mi>âŠ¤</mi></msup><mi>ğ—</mi></mrow><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">\boldsymbol r=\frac{\boldsymbol y^\top\boldsymbol X}{n}</annotation></semantics></math>
is the marginal SNPs-trait inner product and a standardized version,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>ğ«</mi><mo>*</mo></msup><mo>=</mo><mfrac><mrow><msup><mi>ğ²</mi><mrow><mi>*</mi><mi>âŠ¤</mi></mrow></msup><msup><mi>ğ—</mi><mo>*</mo></msup></mrow><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">\boldsymbol r^*=\frac{\boldsymbol y^{*\top}\boldsymbol X^*}{n}</annotation></semantics></math>,
can be recovered from GWAS summary statistics with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>ğ—</mi><mo>*</mo></msup><annotation encoding="application/x-tex">\boldsymbol X^*</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>ğ²</mi><mo>*</mo></msup><annotation encoding="application/x-tex">\boldsymbol y^*</annotation></semantics></math>
being the standardized genotype and phenotype . We note, in this
scenario,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}</annotation></semantics></math>
is only presented for illustration purposes, the actual implementation
does not require the inversion or invertibility of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover><annotation encoding="application/x-tex">\widetilde{\boldsymbol\Sigma}</annotation></semantics></math>;
furthermore, the BRIGHTs method will reduce to LASSOsum when
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Î·</mi><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\eta=0</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mi>Î»</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">p_{\lambda}(\boldsymbol\beta)</annotation></semantics></math>
is chosen as LASSO penalty. The pipeline of the BRIGHTs estimation
procedure with quantitative traits and target minority GWAS summary
statistics is presented in Fig.A, and this pipeline is used for the case
study in Section .</p>
<p>For binary traits, a debiased LASSO estimator,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><mo>=</mo><mover><mi>ğ›</mi><mo accent="true">Ì‚</mo></mover><mo>+</mo><mover><mi>ğš¯</mi><mo accent="true">Ì‚</mo></mover><mrow><mo stretchy="true" form="prefix">(</mo><mi>ğ«</mi><mo>âˆ’</mo><mover><mi>ğ«</mi><mo accent="true">Ì‚</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}=\boldsymbol{\hat b}+\widehat{\boldsymbol\Theta}(\boldsymbol r - \boldsymbol{\hat r})</annotation></semantics></math>,
ensures a valid approximation of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>B</mi><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover></msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="true" form="postfix">|</mo></mrow><mi>ğ›ƒ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)</annotation></semantics></math>
to the negative log-likelihood of logistic regression .
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat b}</annotation></semantics></math>
is the published LASSO estimator from the minority study,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğš¯</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\widehat{\boldsymbol\Theta}</annotation></semantics></math>
is a regularized inverse of the observed information matrix, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>ğ«</mi><mo accent="true">Ì‚</mo></mover><mo>=</mo><mfrac><mrow><mi>e</mi><mi>x</mi><mi>p</mi><mi>i</mi><mi>t</mi><msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>ğ—</mi><mover><mi>ğ›</mi><mo accent="true">Ì‚</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mi>âŠ¤</mi></msup><mi>ğ—</mi></mrow><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">\boldsymbol{\hat r}=\frac{expit(\boldsymbol X\boldsymbol{\hat b})^\top\boldsymbol X}{n}</annotation></semantics></math>
is the marginal marker-predicted outcome inner product, where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>e</mi><mi>x</mi><mi>p</mi><mi>i</mi><mi>t</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>â‹…</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">expit(\cdot)</annotation></semantics></math>
is the inverse of logit function. We note the actual implementation does
not require the debiasing of the LASSO estimator or obtaining
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğš¯</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\widehat{\boldsymbol\Theta}</annotation></semantics></math>,
only
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat b}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ«</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat r}</annotation></semantics></math>
are required in addition to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover><annotation encoding="application/x-tex">\widetilde{\boldsymbol\Sigma}</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>ğ«</mi><annotation encoding="application/x-tex">\boldsymbol r</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">Ìƒ</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\tilde\beta}</annotation></semantics></math>.</p>
<p>We note that for BRIGHTs,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">ÌŒ</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\check\beta}</annotation></semantics></math>
are only required to contain a subset of markers from the LD reference,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğšº</mi><mo accent="true">Ìƒ</mo></mover><annotation encoding="application/x-tex">\widetilde{\boldsymbol\Sigma}</annotation></semantics></math>,
and no further assumptions are made on the covariate space of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat\beta}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›ƒ</mi><mo accent="true">ÌŒ</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\check\beta}</annotation></semantics></math>.</p>
<p>BRIGHTs group of methods utilize a wide variety of summary-level data
from different populations to carry out transfer-learning. We accounted
for Linkage Disequilibrium (LD) via a reference panel (1000 genome
project as default). The reference panel is assumed to be in PLINK 1 <a href="https://www.cog-genomics.org/plink/1.9/input#bed" class="external-link">format</a>.
Summary statistics are expected to be loaded into memory as a
data.frame/data.table.</p>
<p>Below we discuss the required data and implementation tutorials
separately for quantitative traits and binary traits.</p>
</div>
<div class="section level2">
<h2 id="brights-with-quantitative-traits">BRIGHTs with quantitative traits<a class="anchor" aria-label="anchor" href="#brights-with-quantitative-traits"></a>
</h2>
<p>For quantitative traits, BRIGHTs requires the GWAS summary statistics
or marginal genotype-trait inner product,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><msup><mi>ğ—</mi><mi>âŠ¤</mi></msup><mi>ğ²</mi></mrow><mi>n</mi></mfrac><annotation encoding="application/x-tex">\frac{\boldsymbol X^\top\boldsymbol y}{n}</annotation></semantics></math>,
from the target minority population, while from the prior majority
populations either GWAS summary statistics, marginal genotype-trait
inner product, or coefficients estimated from joint models (e.g.Â PRS or
LASSO regression) can be used for model fitting. We note that more than
1 prior majority data can be incorporated in the BRIGHTs model.</p>
<p>First we read the minority summary statistics and majority summary
statistics into R, and provide the <code>ref</code> names of the
reference panel. If <code>ref</code> names are provided as â€œEURâ€, â€œAFRâ€,
â€œEASâ€, â€œSASâ€ ,or â€œAMRâ€, then the default 1000 genome project reference
panels will be used; otherwise <code>ref</code> needs to be provided as
a directory to the plink1 format files (.bim, .bed, .fam).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/UM-KevinHe/BRIGHT" class="external-link">BRIGHT</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority GWAS summary statistics file or marginal genotype-trait inner product file###</span></span>
<span></span>
<span><span class="co"># Read in target GWAS</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="st">"GWAS"</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in target marginal genotype-trait inner product</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="st">"IProd"</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read prior majority GWAS summary statistics file, marginal genotype-trait inner product, or joint coefficient estimates, more than 1 prior majority data can be read in###</span></span>
<span></span>
<span><span class="va">Pind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GWAS"</span>,<span class="st">"IProd"</span>,<span class="st">"Coef"</span><span class="op">)</span></span>
<span><span class="va">Pss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_GWAS1.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss1</span><span class="op">)</span></span>
<span><span class="va">Pss2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_IProd2.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss2</span><span class="op">)</span></span>
<span><span class="va">Pss3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef3.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss3</span><span class="op">)</span></span>
<span><span class="va">Pss</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="va">Pss1</span>,<span class="st">"2"</span><span class="op">=</span><span class="va">Pss2</span>,<span class="st">"3"</span><span class="op">=</span><span class="va">Pss3</span><span class="op">)</span> <span class="co"># The order of list Pss need to be matched with Pind</span></span>
<span></span>
<span><span class="co">### Specify the PLINK file stub of the reference panel or "EUR", "AFR", "EAS", "SAS" ,or "AMR" ###</span></span>
<span><span class="va">ref.bfile</span> <span class="op">&lt;-</span> <span class="st">"refpanel"</span></span>
<span></span>
<span><span class="co">### Read LD region file, only required if ref.bfile is provided as PLINK1 format ###</span></span>
<span><span class="va">LDblocks</span> <span class="op">&lt;-</span> <span class="st">"AFR.hg19"</span> <span class="co"># This will use LD regions as defined in Berisa and Pickrell (2015) for the African population and the hg19 genome build.</span></span>
<span><span class="co"># Other alternatives available. Type ?BRIGHTs for more details. </span></span></code></pre></div>
<p>Reference: <a href="https://academic.oup.com/bioinformatics/article/32/2/283/1743626/Approximately-independent-linkage-disequilibrium" class="external-link">Berisa
and Pickrell (2015)</a></p>
<p>Then, a preprocessing step is required to remove the SNPs that are
not in the reference panel from all data, convert target data into
marginal SNPs-trait inner product, convert prior data into joint
coefficient estimates, and match the effect alleles between the
reference panel and data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">PreprocessS</span><span class="op">(</span>Tss <span class="op">=</span> <span class="va">Tss</span>, Tind <span class="op">=</span> <span class="va">Tind</span>, Pss <span class="op">=</span> <span class="va">Pss</span>, Pind <span class="op">=</span> <span class="va">Pind</span>, ref.bfile<span class="op">=</span><span class="va">ref.bfile</span>, LDblocks<span class="op">=</span><span class="va">LDblocks</span><span class="op">)</span></span></code></pre></div>
<p>Running BRIGHTs using standard pipeline with LASSO penalty on
quantitative traits:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">BRIGHTs</span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>, type.trait<span class="op">=</span><span class="st">"quantitative"</span>, penalty<span class="op">=</span><span class="st">"LASSO"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="brights-with-binary-traits">BRIGHTs with binary traits<a class="anchor" aria-label="anchor" href="#brights-with-binary-traits"></a>
</h2>
<p>For binary traits, in addition to the GWAS summary statistics or
marginal genotype-trait inner product,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><msup><mi>ğ—</mi><mi>âŠ¤</mi></msup><mi>ğ²</mi></mrow><mi>n</mi></mfrac><annotation encoding="application/x-tex">\frac{\boldsymbol X^\top\boldsymbol y}{n}</annotation></semantics></math>,
BRIGHTs requires an estimate based on logistic LASSO regression,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>ğ›</mi><mo accent="true">Ì‚</mo></mover><annotation encoding="application/x-tex">\boldsymbol{\hat b}</annotation></semantics></math>,
and the marginal genotype-predicted traits inner product,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><msup><mi>ğ—</mi><mi>âŠ¤</mi></msup><mi>e</mi><mi>x</mi><mi>p</mi><mi>i</mi><mi>t</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>ğ—</mi><mover><mi>ğ›</mi><mo accent="true">Ì‚</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mi>n</mi></mfrac><annotation encoding="application/x-tex">\frac{\boldsymbol X^\top expit(\boldsymbol X \boldsymbol{\hat b})}{n}</annotation></semantics></math>,
from the target minority population. From the prior majority populations
coefficients estimated from joint models (e.g.Â logistic LASSO
regression) is required for model fitting. We note that more than 1
prior majority data can be incorporated in the BRIGHTs model.</p>
<p>First we read the minority summary statistics and majority summary
statistics into R, and provide the <code>ref</code> names of the
reference panel. If <code>ref</code> names are provided as â€œEURâ€, â€œAFRâ€,
â€œEASâ€, â€œSASâ€ ,or â€œAMRâ€, then the default 1000 genome project reference
panels will be used; otherwise <code>ref</code> needs to be provided as
a directory to the plink1 format files (.bim, .bed, .fam).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/UM-KevinHe/BRIGHT" class="external-link">BRIGHT</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority GWAS summary statistics file or marginal genotype-trait inner product file###</span></span>
<span></span>
<span><span class="co"># Read in target GWAS</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GWAS"</span>,<span class="st">"LASSO"</span>,<span class="st">"IProdPred"</span><span class="op">)</span></span>
<span><span class="va">Tss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in target marginal genotype-trait inner product</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IProd"</span>,<span class="st">"LASSO"</span>,<span class="st">"IProdPred"</span><span class="op">)</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority LASSO estimates file and marginal genotype-predicted outcome inner product file###</span></span>
<span><span class="va">bhat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_LASSO.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bhat</span><span class="op">)</span></span>
<span><span class="va">rhat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProdPred.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bhat</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span> <span class="op">=</span> <span class="va">Tss1</span>, <span class="st">"2"</span> <span class="op">=</span> <span class="va">bhat</span>, <span class="st">"3"</span> <span class="op">=</span> <span class="va">rhat</span><span class="op">)</span> <span class="co"># The order of list Tss need to be matched with Tind</span></span>
<span></span>
<span><span class="co">### Read prior majority GWAS summary statistics file, marginal genotype-trait inner product, or joint coefficient estimates, more than 1 prior majority data can be read in###</span></span>
<span></span>
<span><span class="va">Pind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Coef"</span>, <span class="st">"Coef"</span>, <span class="st">"Coef"</span><span class="op">)</span></span>
<span><span class="va">Pss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef1.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss1</span><span class="op">)</span></span>
<span><span class="va">Pss2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef2.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss2</span><span class="op">)</span></span>
<span><span class="va">Pss3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef3.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss3</span><span class="op">)</span></span>
<span><span class="va">Pss</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="va">Pss1</span>,<span class="st">"2"</span><span class="op">=</span><span class="va">Pss2</span>,<span class="st">"3"</span><span class="op">=</span><span class="va">Pss3</span><span class="op">)</span> <span class="co"># The order of list Pss need to be matched with Pind</span></span>
<span></span>
<span><span class="co">### Specify the PLINK file stub of the reference panel or "EUR", "AFR", "EAS", "SAS" ,or "AMR" ###</span></span>
<span><span class="va">ref.bfile</span> <span class="op">&lt;-</span> <span class="st">"refpanel"</span></span>
<span></span>
<span><span class="co">### Read LD region file, only required if ref.bfile is provided as PLINK1 format ###</span></span>
<span><span class="va">LDblocks</span> <span class="op">&lt;-</span> <span class="st">"AFR.hg19"</span> <span class="co"># This will use LD regions as defined in Berisa and Pickrell (2015) for the African population and the hg19 genome build.</span></span>
<span><span class="co"># Other alternatives available. Type ?BRIGHTs for more details. </span></span></code></pre></div>
<p>Then, a preprocessing step is required to remove the SNPs that are
not in the reference panel from all data, convert target data into
marginal SNPs-trait inner product, convert prior data into joint
coefficient estimates, and match the effect alleles between the
reference panel and data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">PreprocessS</span><span class="op">(</span>Tss <span class="op">=</span> <span class="va">Tss</span>, Tind <span class="op">=</span> <span class="va">Tind</span>, Pss <span class="op">=</span> <span class="va">Pss</span>, Pind <span class="op">=</span> <span class="va">Pind</span>, ref.bfile<span class="op">=</span><span class="va">ref.bfile</span>, LDblocks<span class="op">=</span><span class="va">LDblocks</span><span class="op">)</span></span></code></pre></div>
<p>Running BRIGHTs using standard pipeline with LASSO penalty on
quantitative traits:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">BRIGHTs</span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>, type.trait<span class="op">=</span><span class="st">"binary"</span>, penalty<span class="op">=</span><span class="st">"LASSO"</span><span class="op">)</span></span></code></pre></div>
<p>This procedure requires additional and quite stringent summary
statistics from both target and prior data, in genetics studies its
quite common to treat binary outcome as continuous and perform
continuous models on the data; therefore, in the case where the above
additonal summary statistics are not available, the BRIGHTS with
quantitative traits procedure can also be used to analyze the binary
data.</p>
</div>
<div class="section level2">
<h2 id="model-validation-with-individual-level-test-data">Model validation with individual-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-individual-level-test-data"></a>
</h2>
<p>When individual-level test data is available, <code>BRIGHT</code>
package provide automated validation functions and generates evaluation
plots:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in target individual-level data</span></span>
<span><span class="va">Testgeno</span> <span class="op">&lt;-</span> <span class="st">"/path/to/test/plink"</span></span>
<span><span class="va">Testpheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Test_phenotype.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testpheno</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Ind</span><span class="op">(</span><span class="va">out</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="model-validation-with-summary-level-test-data">Model validation with summary-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-summary-level-test-data"></a>
</h2>
<p>When summary test data is available, <code>BRIGHT</code> package
provide automated validation functions for parameter fine-tunning:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in test GWAS</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"GWAS"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in test marginal genotype-trait inner product</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"IProd"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Sum</span><span class="op">(</span><span class="va">Testss</span>, <span class="va">Testind</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
<p>We note that summary level test data is only supported for
quantitative traits</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qinmengge Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
