<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BRIGHTs tutorial • BRIGHT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BRIGHTs tutorial">
<meta property="og:description" content="BRIGHT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BRIGHT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/BRIGHT.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/BRIGHTi.html">BRIGHTi tutorial</a>
    </li>
    <li>
      <a href="../articles/BRIGHTs.html">BRIGHTs tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BRIGHTs tutorial</h1>
                        <h4 data-toc-skip class="author">Qinmengge Li</h4>
            
            <h4 data-toc-skip class="date">5/1/2023</h4>
      
      
      <div class="hidden name"><code>BRIGHTs.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>To integrate summary statistics without using individual-level data, while adjusting for the potential disparities between prior majority and target minority populations, we propose the following penalized objective function, <span class="math display">\[\begin{align}
Q_{BRIGHTs}(\boldsymbol\beta) =&amp; B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta) + \eta B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\check\beta}||\boldsymbol\beta) + p_{\lambda}(\boldsymbol\beta),\label{eq: BRIGHT}
\end{align}\]</span> where <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> and <span class="math inline">\(\boldsymbol{\check\beta}\)</span> are the estimated coefficients obtained from published studies based on minority and majority populations; <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\)</span> is the regularized block-structured LD estimation  from publicly available minority genotypes (e.g. 1000 genome project). <span class="math inline">\(B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)=(\boldsymbol{\hat\beta}-\boldsymbol{\beta})^\top\widetilde{\boldsymbol\Sigma}(\boldsymbol{\hat\beta}-\boldsymbol{\beta})/n\)</span> is the Bregman-divergence between <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> and <span class="math inline">\(\boldsymbol{\beta}\)</span>, serving as an approximation of the negative log-likelihood from the minority data, which is not directly available due to DataSHIELD constraints; <span class="math inline">\(B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\check\beta}||\boldsymbol\beta)=(\boldsymbol{\check\beta}-\boldsymbol{\beta})^\top\widetilde{\boldsymbol\Sigma}(\boldsymbol{\check\beta}-\boldsymbol{\beta})/n\)</span> is the Bregman-divergence between <span class="math inline">\(\boldsymbol{\check\beta}\)</span> and <span class="math inline">\(\boldsymbol{\beta}\)</span>, and is an approximation of the Bregman-divergence in (). The latter approximation is achieved by <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\approx\widehat{\boldsymbol\Sigma}\)</span>, which is widely deemed to be true in genetics studies .</p>
<p>Although BRIGHTs can incorporate general <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> for data integration, only specific <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> can ensure <span class="math inline">\(B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)\)</span> to be a valid approximation of the minority data negative log-likelihood. Below we discuss how to construct <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> from summary-level information in the case of quantitative traits and binary traits.</p>
<p>For quantitative traits, <span class="math inline">\(\boldsymbol{\hat\beta}=\widetilde{\boldsymbol\Sigma}^{-1}\boldsymbol r\)</span> ensures <span class="math inline">\(B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)\)</span> to be a valid approximation of the OLS loss through <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\approx\widehat{\boldsymbol\Sigma}\)</span> , where <span class="math inline">\(\boldsymbol r=\frac{\boldsymbol y^\top\boldsymbol X}{n}\)</span> is the marginal SNPs-trait inner product and a standardized version, <span class="math inline">\(\boldsymbol r^*=\frac{\boldsymbol y^{*\top}\boldsymbol X^*}{n}\)</span>, can be recovered from GWAS summary statistics with <span class="math inline">\(\boldsymbol X^*\)</span> and <span class="math inline">\(\boldsymbol y^*\)</span> being the standardized genotype and phenotype . We note, in this scenario, <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> is only presented for illustration purposes, the actual implementation does not require the inversion or invertibility of <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\)</span>; furthermore, the BRIGHTs method will reduce to LASSOsum  when <span class="math inline">\(\eta=0\)</span> and <span class="math inline">\(p_{\lambda}(\boldsymbol\beta)\)</span> is chosen as LASSO penalty. The pipeline of the BRIGHTs estimation procedure with quantitative traits and target minority GWAS summary statistics is presented in Fig.A, and this pipeline is used for the case study in Section .</p>
<p>For binary traits, a debiased LASSO estimator, <span class="math inline">\(\boldsymbol{\hat\beta}=\boldsymbol{\hat b}+\widehat{\boldsymbol\Theta}(\boldsymbol r - \boldsymbol{\hat r})\)</span>, ensures a valid approximation of <span class="math inline">\(B_{\widetilde{\boldsymbol\Sigma}}(\boldsymbol{\hat\beta}||\boldsymbol\beta)\)</span> to the negative log-likelihood of logistic regression . <span class="math inline">\(\boldsymbol{\hat b}\)</span> is the published LASSO estimator from the minority study, <span class="math inline">\(\widehat{\boldsymbol\Theta}\)</span> is a regularized inverse of the observed information matrix, and <span class="math inline">\(\boldsymbol{\hat r}=\frac{expit(\boldsymbol X\boldsymbol{\hat b})^\top\boldsymbol X}{n}\)</span> is the marginal marker-predicted outcome inner product, where <span class="math inline">\(expit(\cdot)\)</span> is the inverse of logit function. We note the actual implementation does not require the debiasing of the LASSO estimator or obtaining <span class="math inline">\(\widehat{\boldsymbol\Theta}\)</span>, only <span class="math inline">\(\boldsymbol{\hat b}\)</span> and <span class="math inline">\(\boldsymbol{\hat r}\)</span> are required in addition to <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\)</span>, <span class="math inline">\(\boldsymbol r\)</span> and <span class="math inline">\(\boldsymbol{\tilde\beta}\)</span>.</p>
<p>We note that for BRIGHTs, <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> and <span class="math inline">\(\boldsymbol{\check\beta}\)</span> are only required to contain a subset of markers from the LD reference, <span class="math inline">\(\widetilde{\boldsymbol\Sigma}\)</span>, and no further assumptions are made on the covariate space of <span class="math inline">\(\boldsymbol{\hat\beta}\)</span> and <span class="math inline">\(\boldsymbol{\check\beta}\)</span>.</p>
<p>BRIGHTs group of methods utilize a wide variety of summary-level data from different populations to carry out transfer-learning. We accounted for Linkage Disequilibrium (LD) via a reference panel (1000 genome project as default). The reference panel is assumed to be in PLINK 1 <a href="https://www.cog-genomics.org/plink/1.9/input#bed" class="external-link">format</a>. Summary statistics are expected to be loaded into memory as a data.frame/data.table.</p>
<p>Below we discuss the required data and implementation tutorials separately for quantitative traits and binary traits.</p>
</div>
<div class="section level2">
<h2 id="brights-with-quantitative-traits">BRIGHTs with quantitative traits<a class="anchor" aria-label="anchor" href="#brights-with-quantitative-traits"></a>
</h2>
<p>For quantitative traits, BRIGHTs requires the GWAS summary statistics or marginal genotype-trait inner product, <span class="math inline">\(\frac{\boldsymbol X^\top\boldsymbol y}{n}\)</span>, from the target minority population, while from the prior majority populations either GWAS summary statistics, marginal genotype-trait inner product, or coefficients estimated from joint models (e.g. PRS or LASSO regression) can be used for model fitting. We note that more than 1 prior majority data can be incorporated in the BRIGHTs model.</p>
<p>First we read the minority summary statistics and majority summary statistics into R, and provide the <code>ref</code> names of the reference panel. If <code>ref</code> names are provided as “EUR”, “AFR”, “EAS”, “SAS” ,or “AMR”, then the default 1000 genome project reference panels will be used; otherwise <code>ref</code> needs to be provided as a directory to the plink1 format files (.bim, .bed, .fam).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BRIGHT</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority GWAS summary statistics file or marginal genotype-trait inner product file###</span></span>
<span></span>
<span><span class="co"># Read in target GWAS</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="st">"GWAS"</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in target marginal genotype-trait inner product</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="st">"IProd"</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read prior majority GWAS summary statistics file, marginal genotype-trait inner product, or joint coefficient estimates, more than 1 prior majority data can be read in###</span></span>
<span></span>
<span><span class="va">Pind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GWAS"</span>,<span class="st">"IProd"</span>,<span class="st">"Coef"</span><span class="op">)</span></span>
<span><span class="va">Pss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_GWAS1.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss1</span><span class="op">)</span></span>
<span><span class="va">Pss2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_IProd2.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss2</span><span class="op">)</span></span>
<span><span class="va">Pss3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef3.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss3</span><span class="op">)</span></span>
<span><span class="va">Pss</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="va">Pss1</span>,<span class="st">"2"</span><span class="op">=</span><span class="va">Pss2</span>,<span class="st">"3"</span><span class="op">=</span><span class="va">Pss3</span><span class="op">)</span> <span class="co"># The order of list Pss need to be matched with Pind</span></span>
<span></span>
<span><span class="co">### Specify the PLINK file stub of the reference panel or "EUR", "AFR", "EAS", "SAS" ,or "AMR" ###</span></span>
<span><span class="va">ref.bfile</span> <span class="op">&lt;-</span> <span class="st">"refpanel"</span></span>
<span></span>
<span><span class="co">### Read LD region file, only required if ref.bfile is provided as PLINK1 format ###</span></span>
<span><span class="va">LDblocks</span> <span class="op">&lt;-</span> <span class="st">"AFR.hg19"</span> <span class="co"># This will use LD regions as defined in Berisa and Pickrell (2015) for the African population and the hg19 genome build.</span></span>
<span><span class="co"># Other alternatives available. Type ?BRIGHTs for more details. </span></span></code></pre></div>
<p>Reference: <a href="https://academic.oup.com/bioinformatics/article/32/2/283/1743626/Approximately-independent-linkage-disequilibrium" class="external-link">Berisa and Pickrell (2015)</a></p>
<p>Then, a preprocessing step is required to remove the SNPs that are not in the reference panel from all data, convert target data into marginal SNPs-trait inner product, convert prior data into joint coefficient estimates, and match the effect alleles between the reference panel and data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PreprocessS.html">PreprocessS</a></span><span class="op">(</span>Tss <span class="op">=</span> <span class="va">Tss</span>, Tind <span class="op">=</span> <span class="va">Tind</span>, Pss <span class="op">=</span> <span class="va">Pss</span>, Pind <span class="op">=</span> <span class="va">Pind</span>, ref.bfile<span class="op">=</span><span class="va">ref.bfile</span>, LDblocks<span class="op">=</span><span class="va">LDblocks</span><span class="op">)</span></span></code></pre></div>
<p>Running BRIGHTs using standard pipeline with LASSO penalty on quantitative traits:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BRIGHTs.html">BRIGHTs</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>, type.trait<span class="op">=</span><span class="st">"quantitative"</span>, penalty<span class="op">=</span><span class="st">"LASSO"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="brights-with-binary-traits">BRIGHTs with binary traits<a class="anchor" aria-label="anchor" href="#brights-with-binary-traits"></a>
</h2>
<p>For binary traits, in addition to the GWAS summary statistics or marginal genotype-trait inner product, <span class="math inline">\(\frac{\boldsymbol X^\top\boldsymbol y}{n}\)</span>, BRIGHTs requires an estimate based on logistic LASSO regression, <span class="math inline">\(\boldsymbol{\hat b}\)</span>, and the marginal genotype-predicted traits inner product, <span class="math inline">\(\frac{\boldsymbol X^\top expit(\boldsymbol X \boldsymbol{\hat b})}{n}\)</span>, from the target minority population. From the prior majority populations coefficients estimated from joint models (e.g. logistic LASSO regression) is required for model fitting. We note that more than 1 prior majority data can be incorporated in the BRIGHTs model.</p>
<p>First we read the minority summary statistics and majority summary statistics into R, and provide the <code>ref</code> names of the reference panel. If <code>ref</code> names are provided as “EUR”, “AFR”, “EAS”, “SAS” ,or “AMR”, then the default 1000 genome project reference panels will be used; otherwise <code>ref</code> needs to be provided as a directory to the plink1 format files (.bim, .bed, .fam).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BRIGHT</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority GWAS summary statistics file or marginal genotype-trait inner product file###</span></span>
<span></span>
<span><span class="co"># Read in target GWAS</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GWAS"</span>,<span class="st">"LASSO"</span>,<span class="st">"IProdPred"</span><span class="op">)</span></span>
<span><span class="va">Tss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in target marginal genotype-trait inner product</span></span>
<span><span class="va">Tind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IProd"</span>,<span class="st">"LASSO"</span>,<span class="st">"IProdPred"</span><span class="op">)</span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Tss</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read target minority LASSO estimates file and marginal genotype-predicted outcome inner product file###</span></span>
<span><span class="va">bhat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_LASSO.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bhat</span><span class="op">)</span></span>
<span><span class="va">rhat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProdPred.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bhat</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span> <span class="op">=</span> <span class="va">Tss1</span>, <span class="st">"2"</span> <span class="op">=</span> <span class="va">bhat</span>, <span class="st">"3"</span> <span class="op">=</span> <span class="va">rhat</span><span class="op">)</span> <span class="co"># The order of list Tss need to be matched with Tind</span></span>
<span></span>
<span><span class="co">### Read prior majority GWAS summary statistics file, marginal genotype-trait inner product, or joint coefficient estimates, more than 1 prior majority data can be read in###</span></span>
<span></span>
<span><span class="va">Pind</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Coef"</span>, <span class="st">"Coef"</span>, <span class="st">"Coef"</span><span class="op">)</span></span>
<span><span class="va">Pss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef1.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss1</span><span class="op">)</span></span>
<span><span class="va">Pss2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef2.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss2</span><span class="op">)</span></span>
<span><span class="va">Pss3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Prior_Coef3.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Pss3</span><span class="op">)</span></span>
<span><span class="va">Pss</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="va">Pss1</span>,<span class="st">"2"</span><span class="op">=</span><span class="va">Pss2</span>,<span class="st">"3"</span><span class="op">=</span><span class="va">Pss3</span><span class="op">)</span> <span class="co"># The order of list Pss need to be matched with Pind</span></span>
<span></span>
<span><span class="co">### Specify the PLINK file stub of the reference panel or "EUR", "AFR", "EAS", "SAS" ,or "AMR" ###</span></span>
<span><span class="va">ref.bfile</span> <span class="op">&lt;-</span> <span class="st">"refpanel"</span></span>
<span></span>
<span><span class="co">### Read LD region file, only required if ref.bfile is provided as PLINK1 format ###</span></span>
<span><span class="va">LDblocks</span> <span class="op">&lt;-</span> <span class="st">"AFR.hg19"</span> <span class="co"># This will use LD regions as defined in Berisa and Pickrell (2015) for the African population and the hg19 genome build.</span></span>
<span><span class="co"># Other alternatives available. Type ?BRIGHTs for more details. </span></span></code></pre></div>
<p>Then, a preprocessing step is required to remove the SNPs that are not in the reference panel from all data, convert target data into marginal SNPs-trait inner product, convert prior data into joint coefficient estimates, and match the effect alleles between the reference panel and data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PreprocessS.html">PreprocessS</a></span><span class="op">(</span>Tss <span class="op">=</span> <span class="va">Tss</span>, Tind <span class="op">=</span> <span class="va">Tind</span>, Pss <span class="op">=</span> <span class="va">Pss</span>, Pind <span class="op">=</span> <span class="va">Pind</span>, ref.bfile<span class="op">=</span><span class="va">ref.bfile</span>, LDblocks<span class="op">=</span><span class="va">LDblocks</span><span class="op">)</span></span></code></pre></div>
<p>Running BRIGHTs using standard pipeline with LASSO penalty on quantitative traits:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BRIGHTs.html">BRIGHTs</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>, type.trait<span class="op">=</span><span class="st">"binary"</span>, penalty<span class="op">=</span><span class="st">"LASSO"</span><span class="op">)</span></span></code></pre></div>
<p>This procedure requires additional and quite stringent summary statistics from both target and prior data, in genetics studies its quite common to treat binary outcome as continuous and perform continuous models on the data; therefore, in the case where the above additonal summary statistics are not available, the BRIGHTS with quantitative traits procedure can also be used to analyze the binary data.</p>
</div>
<div class="section level2">
<h2 id="model-validation-with-individual-level-test-data">Model validation with individual-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-individual-level-test-data"></a>
</h2>
<p>When individual-level test data is available, <code>BRIGHT</code> package provide automated validation functions and generates evaluation plots:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in target individual-level data</span></span>
<span><span class="va">Testgeno</span> <span class="op">&lt;-</span> <span class="st">"/path/to/test/plink"</span></span>
<span><span class="va">Testpheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Test_phenotype.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testpheno</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Ind</span><span class="op">(</span><span class="va">out</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="model-validation-with-summary-level-test-data">Model validation with summary-level test data<a class="anchor" aria-label="anchor" href="#model-validation-with-summary-level-test-data"></a>
</h2>
<p>When summary test data is available, <code>BRIGHT</code> package provide automated validation functions for parameter fine-tunning:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in test GWAS</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"GWAS"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_GWAS.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively read in test marginal genotype-trait inner product</span></span>
<span><span class="va">Testind</span><span class="op">=</span><span class="st">"IProd"</span></span>
<span><span class="va">Testss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"Target_IProd.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Testss</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform testing</span></span>
<span><span class="va">Val</span> <span class="op">&lt;-</span> <span class="fu">Valid.Sum</span><span class="op">(</span><span class="va">Testss</span>, <span class="va">Testind</span>, <span class="va">Testpheno</span>, <span class="va">Testgeno</span><span class="op">)</span></span></code></pre></div>
<p>We note that summary level test data is only supported for quantitative traits</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Qinmengge Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
