<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Preprocess for summary statistics (BRIGHTs) — PreprocessS • BRIGHT</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Preprocess for summary statistics (BRIGHTs) — PreprocessS"><meta property="og:description" content='Preprocess the GWAS and the reference. Quality control is performed first for both target and prior (please see QC section). Then prior "GWAS" or "Correlations" is converted to "Coefficients" through BRIGHTs with eta=0; the hyperparameters are tuned through BRCp.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BRIGHT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/BRIGHT.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/BRIGHTi.html">BRIGHTi tutorial</a>
    </li>
    <li>
      <a href="../articles/BRIGHTs.html">BRIGHTs tutorial</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/UM-KevinHe/BRIGHT/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Preprocess for summary statistics (BRIGHTs)</h1>
    <small class="dont-index">Source: <a href="https://github.com/UM-KevinHe/BRIGHT/blob/HEAD/R/PreprocessS.R" class="external-link"><code>R/PreprocessS.R</code></a></small>
    <div class="hidden name"><code>PreprocessS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Preprocess the GWAS and the reference. Quality control is performed first for both target and prior (please see QC section). Then prior "GWAS" or "Correlations" is converted to "Coefficients" through BRIGHTs with eta=0; the hyperparameters are tuned through BRCp.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">PreprocessS</span><span class="op">(</span></span>
<span>  <span class="va">Tss</span>,</span>
<span>  <span class="va">Tind</span>,</span>
<span>  <span class="va">Tref</span>,</span>
<span>  Pss <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  Pind <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  Pref <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  TLDblocks <span class="op">=</span> <span class="st">"EUR.hg19"</span>,</span>
<span>  PLDblocks <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  TN <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  PN <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>Tss</dt>
<dd><p>a data frame of the target summary statistics, including "GWAS" or "Corr" corresponding to the GWAS or correlation summary statistics, the input summary type must match with Tind below.</p></dd>


<dt>Tind</dt>
<dd><p>a character scaler of the input target summary type, it must be either "GWAS" or "Corr" corresponding to the GWAS or correlation summary statistics. It must match with the summary statistics in Tss.</p></dd>


<dt>Tref</dt>
<dd><p>a character scaler indicating the LD reference. It takes values as either a directory to user provided Plink 1 binary files (.bim, .bed, .fam) or the 1000 genome project reference provided by the package: "EUR.hg19", "EUR.hg38", "AFR.hg19", "AFR.hg38", "EAS.hg19", "EAS.hg38", "AMR.hg19", "AMR.hg38", "SAS.hg19" or "SAS.hg38". The abbreviations before dot are populations (more detials can be found here: https://useast.ensembl.org/Help/Faq?id=532); The abbreviations after dot are the genetic build (more details can be found: https://gatk.broadinstitute.org/hc/en-us/articles/360035890951-Human-genome-reference-builds-GRCh38-or-hg38-b37-hg19). For the user provided reference, the SNPs must match with those provided in Pref with the same order.</p></dd>


<dt>Pss</dt>
<dd><p>a data frame of the prior summary statistics, including "GWAS", "Corr" or "Coef" corresponding to the GWAS, correlation or coefficient summary statistics, the input summary type must match with Pind below.</p></dd>


<dt>Pind</dt>
<dd><p>a character scaler of the input prior summary type, it must be either "GWAS", "Corr" or "Coef" corresponding to the GWAS, correlation or coefficient summary statistics. It must match with the summary statistics in Pss.</p></dd>


<dt>Pref</dt>
<dd><p>a character scaler indicating the LD reference. It takes values as either a directory to user provided Plink 1 binary files (.bim, .bed, .fam) or the 1000 genome project reference provided by the package: "EUR.hg19", "EUR.hg38", "AFR.hg19", "AFR.hg38", "EAS.hg19", "EAS.hg38", "AMR.hg19", "AMR.hg38", "SAS.hg19" or "SAS.hg38". The abbreviations before dot are populations (more detials can be found here: https://useast.ensembl.org/Help/Faq?id=532); The abbreviations after dot are the genetic build (more details can be found: https://gatk.broadinstitute.org/hc/en-us/articles/360035890951-Human-genome-reference-builds-GRCh38-or-hg38-b37-hg19). For the user provided reference, the SNPs must match with those provided in Tref with the same order.</p></dd>


<dt>TLDblocks</dt>
<dd><p>a character scaler indicating the LD block structure to be used from Berisa et al for the target population. It takes values from "EUR.hg19", "EUR.hg38", "AFR.hg19", "AFR.hg38", "EAS.hg19", "EAS.hg38", "AMR.hg19", "AMR.hg38", "SAS.hg19" or "SAS.hg38". The abbreviations before dot are populations (more detials can be found here: https://useast.ensembl.org/Help/Faq?id=532); The abbreviations after dot are the genetic build (more details can be found: https://gatk.broadinstitute.org/hc/en-us/articles/360035890951-Human-genome-reference-builds-GRCh38-or-hg38-b37-hg19). The genetic build must match with those provided in Tref.</p></dd>


<dt>PLDblocks</dt>
<dd><p>a character scaler indicating the LD block structure to be used from Berisa et al for the prior population. It takes values from "EUR.hg19", "EUR.hg38", "AFR.hg19", "AFR.hg38", "EAS.hg19", "EAS.hg38", "AMR.hg19", "AMR.hg38", "SAS.hg19" or "SAS.hg38". The abbreviations before dot are populations (more detials can be found here: https://useast.ensembl.org/Help/Faq?id=532); The abbreviations after dot are the genetic build (more details can be found: https://gatk.broadinstitute.org/hc/en-us/articles/360035890951-Human-genome-reference-builds-GRCh38-or-hg38-b37-hg19). The genetic build must match with those provided in Pref.</p></dd>


<dt>TN</dt>
<dd><p>a integer scaler indicating the sample sizes of the target summary statistics input in Tss.</p></dd>


<dt>PN</dt>
<dd><p>a integer scaler indicating the sample sizes of the target summary statistics input in Tss.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of preprocessed data that can be directly input to BRIGHTs().</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Qinmengge Li.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

